Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.cpp

```diff
@@ -0,0 +1,515 @@
+/*!
+  \file NodeWiseMeshPartitioner.cpp
+  \date   2016.05
+
+  \brief  Define the members of class NodeWiseMeshPartitioner
+
+  \copyright
+  Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+             Distributed under a Modified BSD License.
+               See accompanying file LICENSE.txt or
+               http://www.opengeosys.org/project/license
+
+*/
+
+#include "NodeWiseMeshPartitioner.h"
+
+#include <limits>
+#include <iomanip>
+#include <cstdio>  // for binary output
+
+#include <logog/include/logog.hpp>
+
+#include "BaseLib/Error.h"
+
+#include "MeshLib/IO/VtkIO/VtuInterface.h"
+
+#include "MeshLib/Node.h"
+#include "MeshLib/Elements/Element.h"
+
+namespace ApplicationUtils
+{
+struct NodeStruct
+{
+    NodeWiseMeshPartitioner::IntegerType id;
+    double x;
+    double y;
+    double z;
+};
+
+void NodeWiseMeshPartitioner::readMetisData(const std::string& file_name_base)
+{
+    const std::string npartitions_str = std::to_string(_npartitions);
+
+    // Read partitioned mesh data from METIS
+    std::string fname_parts = file_name_base + ".mesh.npart." + npartitions_str;
+
+    std::ifstream npart_in(fname_parts.data());
+    if (!npart_in.is_open())
+    {
+        OGS_FATAL(
+            "Error: cannot open file %s. It may not exist! \
+                   \n Run mpmetis beforehand or use option -m",
+            fname_parts.data());
+    }
+
+    const std::size_t nnodes = _mesh->getNumberOfNodes();
+
+    std::size_t counter = 0;
+    while (!npart_in.eof())
+    {
+        npart_in >> _nodes_partition_ids[counter++] >> std::ws;
+        if (counter == nnodes)
+          break;
+    }
+
+    npart_in.close();
+
+    if( counter != nnodes)
+    {
+        OGS_FATAL(
+            "Error: data in %s are less than expected.", fname_parts.data());
+    }
+
+    // TEST  std::remove(fname_parts.c_str());
+}
+
+void NodeWiseMeshPartitioner::partitionByMETIS(const bool is_mixed_hl_elem)
+{
+    std::vector<MeshLib::Node*> const& nodes = _mesh->getNodes();
+    for (std::size_t part_id = 0; part_id < _partitions.size(); part_id++)
+    {
+        auto& partition = _partitions[part_id];
+
+        INFO("Processing partition: %d", part_id);
+
+        // Find non-ghost nodes in this partition
+        // -- Extra nodes for high order elements
+        std::vector<MeshLib::Node*> extra_nodes;
+        for (std::size_t i = 0; i < _mesh->getNumberOfNodes(); i++)
+        {
+            if (_nodes_partition_ids[i] == part_id)
+            {
+                if (is_mixed_hl_elem)
+                {
+                    if (i < _mesh->getNumberOfBaseNodes())
+                        partition.nodes.push_back(nodes[i]);
+                    else
+                        extra_nodes.push_back(nodes[i]);
+                }
+                {
+                    partition.nodes.push_back(nodes[i]);
+                }
```

Comment:
Yes, ``else`` is missing. Thanks for pointing out.  Since there is not such example (is_mixed_hl_elem) so far, the content under if-condition is not tested (though it is straightforward and simple), and I added a comment as
```
  if (is_mixed_hl_elem)  // TODO: Test it once there is a case
```